src/parquet/arrow/arrow-reader-writer-test.cc

```diff
@@ -1104,61 +1103,300 @@ class TestNestedSchemaRead : public ::testing::Test {
+    // this needs to change when a nested writes are implemented
 
     // create the schema:
-    // required group group1 {
+    // <struct_repetition> group group1 {
     //   required int32 leaf1;
-    //   required int32 leaf2;
+    //   optional int32 leaf2;
     // }
     // required int32 leaf3;
 
-    parquet_fields.push_back(GroupNode::Make("group1", Repetition::REQUIRED,
+    parquet_fields.push_back(GroupNode::Make("group1", struct_repetition,
         {PrimitiveNode::Make("leaf1", Repetition::REQUIRED, ParquetType::INT32),
-            PrimitiveNode::Make("leaf2", Repetition::REQUIRED, ParquetType::INT32)}));
+         PrimitiveNode::Make("leaf2", Repetition::OPTIONAL, ParquetType::INT32)}));
     parquet_fields.push_back(
         PrimitiveNode::Make("leaf3", Repetition::REQUIRED, ParquetType::INT32));
 
-    const int num_columns = 3;
     auto schema_node = GroupNode::Make("schema", Repetition::REQUIRED, parquet_fields);
 
-    InitNewParquetFile(std::static_pointer_cast<GroupNode>(schema_node), 0);
+    // Create definition levels for the different columns that contain interleaved
+    // nulls and values at all nesting levels
+
+    //  definition levels for optional fields
+    int16_t leaf1_def_levels[NUM_SIMPLE_TEST_ROWS];
+    int16_t leaf2_def_levels[NUM_SIMPLE_TEST_ROWS];
+    int16_t leaf3_def_levels[NUM_SIMPLE_TEST_ROWS];
+    for (int i = 0; i < NUM_SIMPLE_TEST_ROWS; i++)  {
+      // leaf1 is required within the optional group1, so it is only null
+      // when the group is null
+      leaf1_def_levels[i] = (i % 3 == 0) ? 0 : 1;
+      // leaf2 is optional, can be null in the primitive (def-level 1) or
+      // struct level (def-level 0)
+      leaf2_def_levels[i] = i % 3;
+      // leaf3 is required
+      leaf3_def_levels[i] = 0;
+    }
+
+    int16_t rep_levels[NUM_SIMPLE_TEST_ROWS];
+    memset(rep_levels, 0, sizeof(rep_levels));
 
-    for (int i = 0; i < num_columns; i++) {
-      auto column_writer = row_group_writer_->NextColumn();
-      auto typed_writer = reinterpret_cast<TypedColumnWriter<Int32Type>*>(column_writer);
-      typed_writer->WriteBatch(0, nullptr, nullptr, nullptr);
+    // Produce values for the columns
+    MakeValues(NUM_SIMPLE_TEST_ROWS);
+    int32_t* values = reinterpret_cast<int32_t*>(values_array_->data()->mutable_data());
+
+    // Create the actual parquet file
+    InitNewParquetFile(std::static_pointer_cast<GroupNode>(schema_node),
+      NUM_SIMPLE_TEST_ROWS);
+
+    // leaf1 column
+    WriteColumnData(NUM_SIMPLE_TEST_ROWS, leaf1_def_levels, rep_levels, values);
+    // leaf2 column
+    WriteColumnData(NUM_SIMPLE_TEST_ROWS, leaf2_def_levels, rep_levels, values);
+    // leaf3 column
+    WriteColumnData(NUM_SIMPLE_TEST_ROWS, leaf3_def_levels, rep_levels, values);
+
+    FinalizeParquetFile();
+    InitReader();
+  }
+
+  NodePtr CreateSingleTypedNestedGroup(int index, int depth, int num_children,
+      Repetition::type node_repetition, ParquetType::type leaf_type) {
```

Comment:
I was thinking something like `template <typename primitive_type>` with `std::enable_if`
But I can accept this in test code. Leave comments to other reviewers..